### ISSUE TITLE
Converter using primitive types fails when used in UDT array in PostgreSQL

### ISSUE THREAD

**Issue by axelfontaine** [jOOQ/jOOQ#10808]
The following converter:

```kotlin
class HeartRateConverter : Converter<Short, HeartRate> {
    override fun from(databaseObject: Short?): HeartRate? =
        if (databaseObject == null) {
            null
        } else {
            HeartRate(databaseObject.toShort())
        }

    override fun to(userObject: HeartRate?): Short? =
        if (userObject == null) {
            null
        } else {
            java.lang.Short.valueOf(userObject.heartRate) as Short
        }

    override fun fromType(): Class<Short> = Short::class.java
    override fun toType(): Class<HeartRate> = HeartRate::class.java
}
```

fails at runtime with

```
java.lang.UnsupportedOperationException: Class short is not supported
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3521) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3436) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3518) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgSetValue(DefaultBinding.java:3570) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.access$200(DefaultBinding.java:3309) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding$1.operate(DefaultBinding.java:3562) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:131) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgNewRecord(DefaultBinding.java:3555) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:3390) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:3309) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:863) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.pgGetArray(DefaultBinding.java:1204) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1142) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:996) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:863) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1715) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1674) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1639) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:131) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1603) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1578) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:350) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:336) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:243) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:308) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:375) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:333) ~[jooq-3.14.1.jar:na]
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2880) ~[jooq-3.14.1.jar:na]
```

Replacing `Short` with `java.lang.Short` and annotating the class with `@Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")` works around the issue.

The same behavior can be observed with other primitive wrapper types like `java.lang.Long` and `java.lang.Integer`.

### Versions

- jOOQ: 3.14.1
- Java: 14

**Comment by lukaseder** [comment]
Thanks a lot for your report.

It looks like you're using array types? Can you please show an example of a `CREATE TABLE` statement, and the relevant parts of your code generator configuration showing how you applied this converter?

**Comment by lukaseder** [comment]
(The bug is obviously in that this `pgFromString()` method doesn't support primitive types, such as `short.class`, only wrapper types like `Short.class`. But I'm curious about a test case to help reproduce this...

**Comment by lukaseder** [comment]
@axelfontaine, this might still fit in today's planned 3.14.2 release, but I would need a `CREATE TABLE` statement and the code generator config for a reproducer, if you have time...

**Comment by axelfontaine** [comment]
Sorry for the delay. Yes, I (finally) have a little bit of time today. Let me try and extract the relevant bits:

```sql
CREATE TYPE xyz AS
(
    x INT,
    y  INT,
    z  SMALLINT
);

CREATE TABLE a
(
    id       BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    b   xyz[] NOT NULL
);
```

```kotlin
org.jooq.meta.jaxb.ForcedType()
    .withIncludeExpression(".*\\..*\\.z")
    .withIncludeTypes("smallint")
    .withUserType("com.veturanto.server.track.HeartRate")
    .withConverter("com.veturanto.server.database.HeartRateConverter"),
```

As noted `Integer` and `Long` exhibit the same issue.

**Comment by lukaseder** [comment]
Thanks a lot. I'll have to try to reproduce this next week.

**Comment by lukaseder** [comment]
I wonder if your `Converter` specification is really correct, though. The SQL type is `Short?`, not `Short`. Curiously there's no compilation error in your code, perhaps for interop reasons? Does this fix the problem?

```kotlin
public class HeartRateConverter : Converter<Short?, HeartRate> {
    override fun from(databaseObject: Short?): HeartRate? =
            if (databaseObject == null) {
                null
            } else {
                HeartRate(databaseObject.toShort())
            }

    override fun to(userObject: HeartRate?): Short? =
            if (userObject == null) {
                null
            } else {
                java.lang.Short.valueOf(userObject.heartRate) as Short
            }

    // Alternatively, use the "kotlin way" of creating Class<Short?>, which I 
    // couldn't find easily. This is not sufficient: https://stackoverflow.com/a/52489988/521799
    override fun fromType(): Class<Short?> = SQLDataType.SMALLINT.type
    override fun toType(): Class<HeartRate> = HeartRate::class.java
}
```

**Comment by lukaseder** [comment]
I can reproduce this with Java:

```sql
CREATE TYPE u_10808 AS (
  a INT,
  b INT
);

CREATE TABLE t_10808 (
  id INT NOT NULL,
  u u_10808[],
  
  CONSTRAINT pk_t_10808 PRIMARY KEY (id)
);
```

And then:

```java
package org.jooq.test.all.converters;

import org.jooq.impl.AbstractConverter;

public class IntToLongConverter extends AbstractConverter<Integer, Long> {

    /**
     * Generated UID
     */
    private static final long serialVersionUID = -855157555996030997L;

    public IntToLongConverter() {
        super(int.class, long.class);
    }

    @Override
    public Long from(Integer t) {
        return t == null ? null : t.longValue();
    }

    @Override
    public Integer to(Long u) {
        return u == null ? null : u.intValue();
    }
}
```

Finally:

```java
package org.jooq.test.all.converters;

import org.jooq.impl.AbstractConverter;

public class IntToLongConverter extends AbstractConverter<Integer, Long> {

    /**
     * Generated UID
     */
    private static final long serialVersionUID = -855157555996030997L;

    public IntToLongConverter() {
        super(int.class, long.class);
    }

    @Override
    public Long from(Integer t) {
        return t == null ? null : t.longValue();
    }

    @Override
    public Integer to(Long u) {
        return u == null ? null : u.intValue();
    }
}
```

Producing:

```
20:51:55,388 ERROR [org.jooq.impl.DefaultBinding                      ] - Cannot parse array       : {"(2,3)"}
java.lang.UnsupportedOperationException: Class int is not supported
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3522) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3436) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgFromString(DefaultBinding.java:3519) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgSetValue(DefaultBinding.java:3571) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding$1.operate(DefaultBinding.java:3563) ~[classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:131) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgNewRecord(DefaultBinding.java:3556) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:3390) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:1) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:863) ~[classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.pgGetArray(DefaultBinding.java:1204) [classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1142) [classes/:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1) [classes/:?]
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:863) [classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1715) [classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1674) [classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1) [classes/:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:131) [classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1603) [classes/:?]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1578) [classes/:?]
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:350) [classes/:?]
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:336) [classes/:?]
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:243) [classes/:?]
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:308) [classes/:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:375) [classes/:?]
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:333) [classes/:?]
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2880) [classes/:?]
	at org.jooq.test.PostgresTest.testPostgresWithPrimitiveTypeConverter(PostgresTest.java:3606) [test-classes/:?]
```

As I suspected, using `Converter<Short, HeartRate>` seems to be a mistake. It should have been `Converter<Short?, HeartRate>`. Of course, we can (and should) handle this. jOOQ (just like Java) treat the wrapper and primitive type `Class` literals as the same thing when the value is not null: I.e. `Short.class` and `short.class` should mostly produce the same behaviour.

Nevertheless, if you want to allow nulls, then using `Short?` would be better. I'm not sure why Kotlin doesn't produce compilation errors. Perhaps that's an interop thing? E.g. this doesn't compile:

```kotlin
public interface C<A, B> {
    public fun m1(a: A): B
    public fun m2(b: B): A
}

public class I : C<Short, Int> {
    override fun m1(a: Short?): Int? {
        return null;
    }
    override fun m2(a: Int?): Short? {
        return null;
    }
}
```

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.15.0 and 3.14.4 (#10930). Thanks again for your report!
